pkg/health/checker/checker.go

```diff
@@ -5,14 +5,171 @@ import (
 	"time"
 
 	"github.com/square/p2/pkg/health"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/manifest"
+	rcfields "github.com/square/p2/pkg/rc/fields"
 	"github.com/square/p2/pkg/store/consul"
 	"github.com/square/p2/pkg/store/consul/consulutil"
 	"github.com/square/p2/pkg/types"
 	"github.com/square/p2/pkg/util"
 
 	"github.com/hashicorp/consul/api"
+	klabels "k8s.io/kubernetes/pkg/labels"
 )
 
+type HealthServiceChecker interface {
+	WatchPodOnNode(
+		nodename types.NodeName,
+		podID types.PodID,
+		quitCh <-chan struct{},
+	) (chan health.Result, chan error)
+	WatchService(
+		serviceID string,
+		resultCh chan<- map[types.NodeName]health.Result,
+		errCh chan<- error,
+		quitCh <-chan struct{},
+		watchDelay time.Duration,
+		useHealthService bool,
+		status manifest.StatusStanza,
+	)
+	WatchHealth(
+		resultCh chan []*health.Result,
+		errCh chan<- error,
+		quitCh <-chan struct{},
+		jitterWindow time.Duration,
+	)
+	Service(
+		serviceID string,
+		userHealthService bool,
+		status manifest.StatusStanza,
+	) (map[types.NodeName]health.Result, error)
+}
+
+type ResourceClient interface {
+	GetRCIDsForPod(pod types.PodID) ([]rcfields.ID, error)
+}
+
+type ReplicationControllerStore interface {
+	Get(id rcfields.ID) (rcfields.RC, error)
+}
+
+type LabelReader interface {
+	GetMatches(klabels.Selector, labels.Type) ([]labels.Labeled, error)
+}
+
+type healthServiceChecker struct {
+	healthClient   health.client
+	resourceClient ResourceClient
+	consulClient   consulutil.ConsulClient
+	kv             healthKV
+	consulStore    healthStore
+	rcStore        ReplicationControllerStore
+	labelReader	   LabelReader
+}
+
+func NewHealthServiceChecker(hClient health.Client, resourceClient ResourceClient, cClient consulutil.ConsulClient, rcStore ReplicationControllerStore) HealthServiceChecker {
+	return healthServiceChecker{
+		healthClient:   client,
+		resourceClient: resourceClient,
+		consulClient:   cClient,
+		kv:             cClient.KV(),
+		consulStore:    consul.NewConsulStore(cClient),
+		rcStore: rcStore
+	}
+}
+
+func (h healthServiceChecker) WatchPodOnNode(
+	nodename types.NodeName,
+	podID types.PodID,
+	quitCh <-chan struct{},
+) (chan health.Result, chan error) {
+	return consulHealthChecker(h.(healthServiceChecker)).WatchPodOnNode(nodename, podID, quitCh)
```

Comment:
What's the long term fate for this type (`consulHealthChecker`)? What will happen when we move to only health checking via health service? I predicted that this code would introduce a new type that "covered" both consulHealthChecker and healthServiceHealthChecker and we'd use that everywhere. Can you remind me why that's not possible?

Reply:
That's probably the better approach in the long term. That way when consulHealthChecker code can be removed when we move to only the health service. I'll make those changes.